relay_rpc/src/rpc/params.rs

```diff
@@ -0,0 +1,324 @@
+    fn irn_try_from_tag(value: Value, tag: u32) -> Result<Self::Params, ParamsError>;
+}
+
+/// Errors covering API payload parameter conversion issues.
+#[derive(Debug, thiserror::Error)]
+pub enum ParamsError {
+    /// Pairing API serialization/deserialization issues.
+    #[error("Failure serializing/deserializing Sign API parameters: {0}")]
+    Serde(#[from] serde_json::Error),
+    /// Pairing API invalid response tag.
+    #[error("Response tag={0} does not match any of the Sign API methods")]
+    ResponseTag(u32),
+}
+
+/// https://www.jsonrpc.org/specification#response_object
+///
+/// JSON RPC 2.0 response object can either carry success or error data.
+/// Please note, that relay protocol metadata is used to disambiguate the
+/// response data.
+///
+/// For example:
+/// `RelayProtocolHelpers::irn_try_from_tag` is used to deserialize an opaque
+/// response data into the typed parameters.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum ResponseParams {
+    /// A response with a result.
+    #[serde(rename = "result")]
+    Success(Value),
+
+    /// A response for a failed request.
+    #[serde(rename = "error")]
+    Err(Value),
+}
+
+// Convenience macro to de-duplicate implementation for different parameter
+// sets.
+macro_rules! impl_relay_protocol_metadata {
+    ($param_type:ty,$meta:ident) => {
+        paste! {
+            impl RelayProtocolMetadata for $param_type {
+                fn irn_metadata(&self) -> IrnMetadata {
+                    match self {
+                        [<$param_type>]::SessionPropose(_) => session_propose::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionSettle(_) => session_settle::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionRequest(_) => session_request::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionUpdate(_) => session_update::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionDelete(_) => session_delete::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionEvent(_) => session_event::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionExtend(_) => session_extend::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::SessionPing(_) => session_ping::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::PairingDelete(_) => pairing_delete::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::PairingExtend(_) => pairing_extend::[<IRN_ $meta:upper _METADATA>],
+                        [<$param_type>]::PairingPing(_) => pairing_ping::[<IRN_ $meta:upper _METADATA>],
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Convenience macro to de-duplicate implementation for different parameter
+// sets.
+macro_rules! impl_relay_protocol_helpers {
+    ($param_type:ty) => {
+        paste! {
```

Comment:
Is paste! needed here (cannot see its params `[<…>]` inside the call)?

Reply:
you're indeed right! thanks